File: github_clone.ml
(*
  Backup one's public repositories to a directory

  Baris Metin <baris ! metin.org>
*)

open List
open Printf
open Shell
open Json_io
open Json_type.Browse
open Http_client.Convenience

let usage_msg = 
  "\nUsage: github_clone -u USER [-owner] [SYNC_DIRECTORY]\n" ^
  "Backup USER's public github repositories locally.\n" 

let is_owner = ref false
let github_user = ref ""
let sync_dir = ref "."

let optionspeclist = 
  [ 
    ("-owner", Arg.Set (is_owner), "Checkout projects as the owner");
    ("-u", Arg.Set_string (github_user), "Github username");
  ]


(* get name from json object *)
let get_field_from json name =
  hd
    (map (fun (x,y) -> string y)
       (filter (fun (x,y) -> x = name) (objekt json)))


class github user =
  object (self)
    val api_url = "http://github.com/api/v2/json"
    val user = user

    method get url = http_get url
        
    method user =
      let user_url = String.concat "/" [api_url; "user/show"; user] in
      self#get user_url

    method repos =
      let repos_url = String.concat "/" [api_url; "repos/show"; user] in
      self#get repos_url
        
    method url repo = 
      if !is_owner then
        "git@github.com:" ^ user ^ "/" ^ repo ^ ".git"
      else
        "git://github.com/" ^ user ^ "/" ^ repo ^ ".git"

    method repo_names =
      (* return public repositories of a user as a list *)
      let json_obj = json_of_string self#repos in
      let tbl = make_table (objekt json_obj) in
      let lst = array (field tbl "repositories") in
      let rec names lst = 
        match lst with
          [] -> []
        | x::lst' -> (get_field_from x "name") :: (names lst') in
      names lst

    method clone repo target_dir =
      let target_workdir = String.concat "/" [target_dir;repo]
      and repo_url = self#url repo in
      print_endline ("Working on " ^ target_workdir);
      print_endline repo_url;
      if Sys.file_exists target_workdir && Sys.is_directory target_workdir then
        begin
          let cur_dir = Sys.getcwd () in
          Sys.chdir target_workdir;
          call [ cmd "git" ["pull"] ];
          Sys.chdir cur_dir
        end
      else
        begin
          call [ cmd "git" ["clone"; repo_url; target_workdir] ]
        end
        
  end



let () = 
  Arg.parse optionspeclist (fun x -> sync_dir := x) usage_msg;
  
  if !github_user = "" then
    begin
      Arg.usage optionspeclist usage_msg;
      exit 1
    end;

  let g = new github !github_user in
  let rec clone_all repos = 
    match repos with
      [] -> ()
    | repo::repos' ->
        begin
          g#clone repo !sync_dir;
          clone_all repos'
        end
  in

  clone_all g#repo_names

File: github_clone.py
#!/usr/bin/env python

import os
import sys
import json
import urllib
import logging
import optparse
import subprocess

logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s: %(message)s')

class Github(object):
    def __init__(self, options):
        self.api_url = "http://github.com/api/v2/json"
        self.opt = options

    def get(self, url):
        return urllib.urlopen(url).read()

    def user(self):
        return self.get("/".join([self.api_url, "user/show", self.opt.username]))

    def repos(self):
        return self.get("/".join([self.api_url, "repos/show", self.opt.username]))

    def url(self, repo):
        if self.opt.owner:
            return "git@github.com:%s/%s.git" % (self.opt.username, repo)
        else:
            return "git://github.com/%s/%s.git" % (self.opt.username, repo)

    def repo_names(self):
        repos = json.loads(self.repos())
        return [repo["name"] for repo in repos["repositories"]]

    def clone(self, repo, target):
        target_workdir = os.path.join(target, repo)
        repo_url = self.url(repo)
        logging.info("Working on %s" % target_workdir)
        logging.info(repo_url)
        if os.path.isdir(target_workdir):
            cur_dir = os.getcwd()
            os.chdir(target_workdir)
            subprocess.call(["git", "fetch", "origin"])
            subprocess.call(["git", "fetch", "origin", "--tags"])
            subprocess.call(["git", "merge", "master"])
            os.chdir(cur_dir)
        else:
            subprocess.call(["git", "clone", repo_url, target_workdir])


def main():
    parser = optparse.OptionParser()
    parser.add_option(
        "-o", "--owner", dest="owner",
        action="store_true",
        help="Checkout projects as the owner"
        )
    parser.add_option(
        "-u", "--username", dest="username",
        type="string", action="store",
        help="Github username"
        )
    (options, args) = parser.parse_args()
    if not args:
        parser.print_help()
        sys.exit(1)

    target = args[0]
    g = Github(options)
    for repo in g.repo_names():
        g.clone(repo, target)
        

if __name__ == "__main__":
    main()
